File: apps/apple/StillKit/Sources/StillKit/OriginalInstall.swift
import Foundation

/// Which of Apple's two version strings an `OriginalInstallRecord.applicationVersion` holds.
///
/// `AppTransaction.originalAppVersion` is documented as returning `CFBundleVersion` on iOS, iPadOS,
/// tvOS and watchOS, and `CFBundleShortVersionString` on macOS. For Still that means the very same
/// release records a build number like "6" on iPhone and a marketing version like "1.0.1" on Mac.
/// The two are different namespaces and comparing one against the other silently misclassifies a
/// whole platform, so every record states which one it is carrying.
public enum OriginalAppVersionKind: String, Codable, Sendable {
  /// `CFBundleVersion`, Still's build number. What iOS, iPadOS, tvOS and watchOS report.
  case buildNumber
  /// `CFBundleShortVersionString`, Still's marketing version. What macOS reports.
  case marketingVersion
}

/// Local-only facts about when this device first ran a Still build that keeps this record. They are
/// retained in the App Group so an eventual paid-tier change can honor installs from the era when
/// everything was included, even for someone who never created an account. Nothing here is ever
/// sent to Still's backend.
///
/// The record has two halves, because they have very different reliability:
///
///   * The local half is written on first launch from the app's own bundle. It needs no network, no
///     Apple Account and no App Store round trip, so every install gets one.
///   * The verified half comes from Apple's `AppTransaction` and can be absent. It is richer, since
///     it describes the ORIGINAL purchase rather than this device's first launch, but it is only
///     available on newer systems and only when Apple will hand it over. In the sandbox and in
///     TestFlight, Apple documents both of its values as placeholders rather than real data, so a
///     record captured on a test build is not evidence of anything.
///
/// ## Changing this record later, which is the part that can go permanently wrong
///
/// Every record already on a device must keep decoding on every future build. `OriginalInstall`
/// reports a record it cannot decode as no record at all, and the next launch then writes a fresh
/// one dated today, which silently moves the entire existing install base into whatever cohort is
/// current on the day that build ships. That is exactly the permanent misclassification this
/// record exists to prevent, arriving through the back door, and it has already happened once on a
/// development machine while this release was being built. So, without exception:
///
///   * Add a new field as an optional, or give it a default in `init(from:)` below. Never decode a
///     new field with a plain required `decode`, because every record written before it existed
///     would stop decoding.
///   * Never rename or repurpose an existing field. A rename is a removal plus an addition, and
///     the removal is what breaks the old records.
///   * Raise `OriginalInstall.currentSchemaVersion` when the MEANING of a field changes, and read
///     `schemaVersion` before interpreting it. A record with no `schemaVersion` is version 1.
///
/// `OriginalInstallTests` pins all of this: a record in the exact shape this build writes is
/// decoded from raw bytes there, and that test fails the moment a change would make an existing
/// record unreadable.
public struct OriginalInstallRecord: Codable, Equatable, Sendable {
  /// Which shape of this record was written, so a later build can tell what it is reading. Absent
  /// in a record written before this field existed, which is version 1.
  public let schemaVersion: Int

  /// When this record was written: the first launch of a Still build that records it. On a new
  /// install that is the install date. On a device updating from an earlier Still it is the update
  /// date, which is still enough to place someone inside the included-access era. Means exactly the
  /// same thing on every Apple platform.
  public let firstRecordedAt: Date

  /// Still's own marketing version (`CFBundleShortVersionString`) at that first launch. The same
  /// namespace on iOS and macOS, unlike `applicationVersion` below.
  public let firstRecordedAppVersion: String

  /// Apple's verified original application version, or nil when no verified app transaction was
  /// available. Never compare this against a version number without reading
  /// `applicationVersionKind` first: on iOS it is a build number and on macOS it is a marketing
  /// version.
  public let applicationVersion: String?

  /// Which namespace `applicationVersion` is in. Nil when `applicationVersion` is nil, and also
  /// when a later build wrote a namespace this build has never heard of. Either way the answer is
  /// the same: do not compare `applicationVersion` against anything.
  public let applicationVersionKind: OriginalAppVersionKind?

  /// Apple's verified original purchase date: when this Apple Account first obtained Still, on any
  /// device. This one means the same thing on both platforms. Nil when unavailable.
  public let originalPurchaseDate: Date?

  public init(
    firstRecordedAt: Date,
    firstRecordedAppVersion: String,
    applicationVersion: String? = nil,
    applicationVersionKind: OriginalAppVersionKind? = nil,
    originalPurchaseDate: Date? = nil
  ) {
    // Anything this code writes is in the shape this build knows. Only decoding can produce an
    // older one.
    self.schemaVersion = OriginalInstall.currentSchemaVersion
    self.firstRecordedAt = firstRecordedAt
    self.firstRecordedAppVersion = firstRecordedAppVersion
    self.applicationVersion = applicationVersion
    self.applicationVersionKind = applicationVersionKind
    self.originalPurchaseDate = originalPurchaseDate
  }

  /// Written out by hand, and deliberately forgiving, for the reason in this type's documentation:
  /// a record that fails to decode is treated as no record and gets overwritten with today's date.
  /// Fields a later build added are ignored, and an optional field that cannot be understood
  /// becomes nil instead of taking the whole record down with it. Only the two local facts are
  /// required, because a record without them says nothing at all.
  public init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    self.firstRecordedAt = try container.decode(Date.self, forKey: .firstRecordedAt)
    self.firstRecordedAppVersion = try container.decode(
      String.self, forKey: .firstRecordedAppVersion)
    self.schemaVersion = (try? container.decodeIfPresent(Int.self, forKey: .schemaVersion)) ?? 1
    self.applicationVersion = try? container.decodeIfPresent(
      String.self, forKey: .applicationVersion)
    let namespace: String? = try? container.decodeIfPresent(
      String.self, forKey: .applicationVersionKind)
    self.applicationVersionKind = namespace.flatMap { OriginalAppVersionKind(rawValue: $0) }
    self.originalPurchaseDate = try? container.decodeIfPresent(
      Date.self, forKey: .originalPurchaseDate)
  }
}

public enum OriginalInstall {
  static let storageKey = "still:originalInstall"

  /// The shape written today. Raise this only when the meaning of an existing field changes; adding
  /// an optional field does not need a new version. See `OriginalInstallRecord` for the rules that
  /// keep records already on devices readable.
  public static let currentSchemaVersion = 1

  /// How many times a device may ask Apple for its app transaction before giving up for good.
  ///
  /// Asking is not free: Apple returns the locally cached app transaction when there is one and
  /// otherwise requests it from the App Store, which can require the customer to authenticate. On a
  /// device restored from a backup, handed over from someone else, or simply not signed in to the
  /// App Store, an unbounded retry would put a sign-in sheet in front of an app that sells nothing,
  /// at launch, over and over. A small ceiling buys the extra detail where it is available and
  /// stops asking where it is not. The local half of the record survives either way.
  public static let maxVerifiedAttempts = 3

  static let verifiedAttemptsKey = "still:originalInstallVerifiedAttempts"

  /// The namespace `AppTransaction.originalAppVersion` uses on the platform this build runs on.
  /// Derived here, once, so no call site has to remember Apple's platform split.
  public static var applicationVersionKindForThisPlatform: OriginalAppVersionKind {
    #if os(macOS)
      return .marketingVersion
    #else
      return .buildNumber
    #endif
  }

  /// Write the local half exactly once. Read-before-write is the load-bearing behavior: an ordinary
  /// relaunch or an app update must return the existing record, never a fresher one, or an update
  /// would quietly move someone into a newer cohort.
  @discardableResult
  public static func ensure(
    firstRecordedAt: Date,
    appVersion: String,
    defaults: UserDefaults
  ) -> OriginalInstallRecord {
    if let existing = current(defaults) { return existing }
    let record = OriginalInstallRecord(
      firstRecordedAt: firstRecordedAt,
      firstRecordedAppVersion: appVersion
    )
    write(record, defaults: defaults)
    return record
  }

  /// True while it is still worth asking Apple for the verified half: no record yet has it, and the
  /// attempt ceiling has not been reached.
  public static func shouldRequestVerifiedValues(_ defaults: UserDefaults) -> Bool {
    if current(defaults)?.applicationVersion != nil { return false }
    return defaults.integer(forKey: verifiedAttemptsKey) < maxVerifiedAttempts
  }

  /// Count an attempt. Called BEFORE the request, so a request that hangs, fails, or never returns
  /// still moves the device toward the ceiling rather than retrying forever.
  @discardableResult
  public static func countVerifiedAttempt(_ defaults: UserDefaults) -> Int {
    let next = defaults.integer(forKey: verifiedAttemptsKey) + 1
    defaults.set(next, forKey: verifiedAttemptsKey)
    return next
  }

  /// Fill in Apple's verified values, once, leaving the local half exactly as it was written on
  /// first launch. Does nothing when there is no local record yet, or when the verified half is
  /// already present.
  @discardableResult
  public static func fillVerifiedValues(
    applicationVersion: String,
    kind: OriginalAppVersionKind,
    originalPurchaseDate: Date,
    defaults: UserDefaults
  ) -> OriginalInstallRecord? {
    guard let existing = current(defaults) else { return nil }
    guard existing.applicationVersion == nil else { return existing }
    let filled = OriginalInstallRecord(
      firstRecordedAt: existing.firstRecordedAt,
      firstRecordedAppVersion: existing.firstRecordedAppVersion,
      applicationVersion: applicationVersion,
      applicationVersionKind: kind,
      originalPurchaseDate: originalPurchaseDate
    )
    write(filled, defaults: defaults)
    return filled
  }

  /// The stored record, or nil when there is none. Nil also covers a record this build cannot
  /// decode, and `ensure` replaces nil with a record dated today, which is why
  /// `OriginalInstallRecord`'s decoding is forgiving and why its compatibility rules matter.
  public static func current(_ defaults: UserDefaults) -> OriginalInstallRecord? {
    guard let data = defaults.data(forKey: storageKey) else { return nil }
    return try? JSONDecoder().decode(OriginalInstallRecord.self, from: data)
  }

  private static func write(_ record: OriginalInstallRecord, defaults: UserDefaults) {
    if let data = try? JSONEncoder().encode(record) {
      defaults.set(data, forKey: storageKey)
    }
  }
}

File: packages/ext-chromium/lib/original-install.ts
import { browser } from "wxt/browser";

// When this browser first ran a Still build that keeps this record, and which version that was.
//
// It exists so that a later change to what Still includes can recognise the people who installed
// while everything was included, and treat them as promised. Their Apple counterparts are covered
// by the app's own record; a browser has no receipt and no store history to ask, so the browser has
// to write the fact down itself, once, at the moment it first runs.
//
// Two properties are load-bearing:
//
//   * It is local. Nothing here is sent anywhere, no schema changes, and it is readable only by
//     this extension. A cohort record that needed an account would miss exactly the people it
//     exists for, since signing in is optional.
//   * It is written once and never moved forward. An ordinary browser restart, a service worker
//     waking, or an extension update must all return the record already stored. If any of them
//     rewrote it, every existing install would silently re-date itself to the day that build
//     shipped, which is the one mistake this record cannot recover from.
//
// `firstRecordedAt` and `firstRecordedAppVersion` mean exactly what the Apple app's fields of the
// same names mean (`OriginalInstall` in StillKit), so the two surfaces can be compared later
// without anyone having to reconcile two definitions. The Apple record carries an optional richer
// half from Apple's app transaction; there is no browser equivalent, and the local half is the one
// both sides always have.
//
// Changing this record later is the part that can go permanently wrong, for the same reason it is
// on the Apple side. A record that fails to parse reads as no record, and the next start writes a
// fresh one dated today. So: add new fields as optional, never rename or repurpose an existing
// field, and raise `CURRENT_SCHEMA_VERSION` only when the MEANING of a field changes.

const STORAGE_KEY = "still:originalInstall";

/** The shape written today. A stored record with no `schemaVersion` is version 1. */
export const CURRENT_SCHEMA_VERSION = 1;

export interface OriginalInstallRecord {
  /** Which shape of this record was written, so a later build can tell what it is reading. */
  readonly schemaVersion: number;
  /** Milliseconds since the epoch at the first start of a Still build that keeps this record. */
  readonly firstRecordedAt: number;
  /** Still's own extension version at that first start, from the manifest. */
  readonly firstRecordedAppVersion: string;
}

/** The one slot this module reads and writes. Injectable so the tests need no browser. */
export interface OriginalInstallStore {
  get(): Promise<unknown>;
  set(record: OriginalInstallRecord): Promise<void>;
}

export interface OriginalInstallDeps {
  readonly store: OriginalInstallStore;
  /** Milliseconds since the epoch; `Date.now` in real wiring. */
  readonly now: () => number;
  /** This build's extension version. */
  readonly appVersion: string;
}

/**
 * Return the stored record, writing it first if this browser has never had one. Reading before
 * writing is the whole behaviour: it is what makes an update or a worker restart keep the original
 * date instead of replacing it.
 *
 * Never throws. A storage failure costs the record, which is a lost cohort signal for one install
 * and nothing else, and must not be allowed to take down a background start.
 */
export async function ensureOriginalInstall(
  deps: OriginalInstallDeps,
): Promise<OriginalInstallRecord | null> {
  try {
    const existing = parseOriginalInstall(await deps.store.get());
    if (existing) return existing;
    const record: OriginalInstallRecord = {
      schemaVersion: CURRENT_SCHEMA_VERSION,
      firstRecordedAt: deps.now(),
      firstRecordedAppVersion: deps.appVersion,
    };
    await deps.store.set(record);
    return record;
  } catch {
    return null;
  }
}

/**
 * Read a stored value as a record, or null when there is nothing usable there.
 *
 * Deliberately tolerant in one direction only: unknown fields written by a later build are ignored
 * so an older build never destroys a newer record, while a missing or unusable required field
 * reads as absent. A record without a readable first-recorded date says nothing, and pretending
 * otherwise would be worse than rewriting it.
 */
export function parseOriginalInstall(value: unknown): OriginalInstallRecord | null {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return null;
  const record = value as Record<string, unknown>;
  const firstRecordedAt = record.firstRecordedAt;
  const firstRecordedAppVersion = record.firstRecordedAppVersion;
  if (typeof firstRecordedAt !== "number" || !Number.isFinite(firstRecordedAt)) return null;
  if (typeof firstRecordedAppVersion !== "string" || firstRecordedAppVersion.length === 0) {
    return null;
  }
  const schemaVersion = record.schemaVersion;
  return {
    // A record written before the field existed is version 1, which is what it was.
    schemaVersion:
      typeof schemaVersion === "number" && Number.isFinite(schemaVersion)
        ? schemaVersion
        : 1,
    firstRecordedAt,
    firstRecordedAppVersion,
  };
}

/** The production slot, over the same `chrome.storage.local` the rest of the extension uses. */
export function createOriginalInstallStore(): OriginalInstallStore {
  return {
    async get(): Promise<unknown> {
      return (await browser.storage.local.get(STORAGE_KEY))[STORAGE_KEY] ?? null;
    },
    async set(record: OriginalInstallRecord): Promise<void> {
      await browser.storage.local.set({ [STORAGE_KEY]: record });
    },
  };
}
